### ISSUE TITLE
Discussion regarding implementation of  a "sitemap generator" helper

### ISSUE THREAD

**Issue by jaumesala** [mcamara/laravel-localization#48]
Working in one of my projects which has many url, I had to create a method which takes the routes names and generates a simple multilingual sitemap.xml like this:

``` XML
<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
   xmlns:xhtml="http://www.w3.org/1999/xhtml">
      <url>
         <loc>http://example.com</loc>
         <xhtml:link rel="alternate" hreflang="es" href="http://www.example.com" />
         <xhtml:link rel="alternate" hreflang="en" href="http://www.example.com/en" />
      </url>
      <url>
         <loc>http://example.com/servicios</loc>
         <xhtml:link rel="alternate" hreflang="es" href="http://www.example.com/servicios" />
         <xhtml:link rel="alternate" hreflang="en" href="http://www.example.com/en/services" />
      </url>
      ...
</urlset>
```

As you can see, the implementation is really basic, there's no "lastmod", "changefreq"  nor "priority" in the urls.

I would like to implement a helper method in laravel-localization to generate sitemaps. I think it is something that most of us would be pleased to have. If we are really taking so much care in localizing all the routes (with the time that implies), a helper to automatically generate a sitemap would be delightful! Don't you think?

Here is a list of things that I thought this "helper" should be aware of:
-   Check your supportedLocales
-   Check if hideDefaultLanguageInRoute is true 
-   Option to supply extra urls ( i.ex: non localized urls)
-   Option to remove urls (i.ex. exclude all routes that star with "admin" )

My questions are:
-   Should it generate the "priority", "changefreq" and "lastmod" tags?
-   Any ideas about how to extract/store this information?
-   Maybe it's a good idea to store a copy of the sitemap in cache?

Please feel free to give your opinion and recommendations?

**Comment by dhiman252** [comment]
Hi Guys,

Is there any update on the sitemap package?
I am using  the laravel v5.2 and really don't know how to get the list of all the routs in all the different languages. any help would be appreciated.

**Comment by nmpribeiro** [comment]
Let's do it?

**Comment by jasperjorna** [comment]
It's relatively easy to implement with a separate package like this one: https://github.com/Laravelium/laravel-sitemap

I'm basically just looping through the routes and create an entry in the sitemap for each of them.
```php
foreach (array_keys(trans('routes')) as $route) {
    foreach (LaravelLocalization::getSupportedLocales() as $code => $options) {
        // Our default locale is 'en', so we skip it to set 'x-default' below.
        if ($code !== 'en') {
            $translations[$code] = [
                'language' => $code,
                'url'      => LaravelLocalization::getURLFromRouteNameTranslated($code, 'routes.'.$route),
            ];
        } else {
            // Set 'x-default' for our default route to satisfy search engines.
            $translations[$code] = [
                'language' => 'x-default',
                'url'      => LaravelLocalization::getURLFromRouteNameTranslated($code, 'routes.'.$route),
            ];
        }
    }

    foreach ($translations as $translation) {
        $sitemap->add($translation['url'], null, '0.7', 'hourly', [], null, $translations);
    }
}

```

**Comment by jaumesala** [comment]
Hi guys, as I started this thread a few years ago, here is the code I use for my projects. It is not very clean, but I've never had time to improve it (since 2014 hehe). This code is just to give you some ideas to start tinkering. 
Basically, the sitemapRepo filters all the routes in $excludedRoutes and $excludedPartials. Then I generate a multidimensional array to link each page with their alternate. And finally I send this array to the view.

```php
// app/routes/web.php
Route::get('sitemap.xml', ['uses' => 'SitemapController@index', 'as' => 'sitemap']);
```
```php
// app/Http/Controllers/SitemapController.php
namespace App\Http\Controllers;

use Illuminate\Http\Request;
use App\Repositories\SitemapRepository;

class SitemapController extends Controller
{
    protected $sitemapRepo;

    public function __construct(SitemapRepository $sitemapRepo)
    {
        $this->sitemapRepo = $sitemapRepo;
    }

    public function index(Request $request)
    {
        $excludedRoutes = [
            "",
            "cookies",
            "sitemap",
            "legality",
            "privacy",
            "login",
            "logout",
            "register",
        ];

        $excludedPartials = [
            "orders.",
            "password.",
            "admin.",
        ];

        $routes = $this->sitemapRepo->getStaticRoutes($excludedRoutes, $excludedPartials, 1.0);
        
        /* Here you can add custom parsers in case you want to generate 
         * routes for posts, products, profiles, etc. i.e. here is a custom 
         * method that should return the routes of all the posts in your web.
         */

        // $postsRoutes = $this->getPostRoutes('posts.show', 0.9);

        $routes = array_merge($routes, $postsRoutes);
        
        $data = compact('routes');        

        return response()
            ->view('sitemap.xml', $data)
            ->header('Content-Type', 'text/xml');
    }
}
```

```php
// app/Repositories/SitemapRepository.php
namespace App\Repositories;

use Route;
use LaravelLocalization;

class SitemapRepository
{

    public function getStaticRoutes($excludedRoutes = [], $excludedPartials = [], $priority = 1.0)
    {
        $routes = $this->getStaticRouteNames($excludedRoutes, $excludedPartials);
        $locales = LaravelLocalization::getSupportedLanguagesKeys();
        $staticRoutes = [];

        foreach( $locales as $locale ){

            foreach ( $routes as $route){
                $r = [];
                $loc = $this->getLocalizedURL( $locale, $route);
                $r['loc'] = $loc ? $loc : "";
                $r['priority'] = $priority;

                foreach ($locales as $alternate) {
                    $alt = $this->getLocalizedURL( $alternate, $route);
                    $r['alternates'][$alternate] =  $alt ? $alt : "";
                }
                $staticRoutes[] = $r;
            }
        }
        return $staticRoutes;
    }

    public function getStaticRouteNames($excludedRoutes = [], $excludedPartials = [])
    {
        $allRoutes = Route::getRoutes();

        $staticRouteNames = array();

        foreach ( $allRoutes as $route){

            if(in_array("GET", $route->methods))
            {
                if( ! in_array($route->getName(), $excludedRoutes )
                    && ! $this->strpos_array($route->getName(), $excludedPartials )
                    )
                {
                    $staticRouteNames [] = $route->getname();
                }
            }
        }

        return $staticRouteNames;
    }

    private function strpos_array($haystack, $needle, $offset=0)
    {
        if(!is_array($needle)) $needle = array($needle);

        foreach($needle as $query) {
            if(strpos($haystack, $query, $offset) !== false) return true; // stop on first true result
        }

        return false;
    }

    public function getLocalizedURL($locale, $baseRouteName, $args = [])
    {
        return LaravelLocalization::getURLFromRouteNameTranslated( $locale, 'routes.'.$baseRouteName, $args);
    }

    public function getSupportedLocalesKeys()
    {
        return LaravelLocalization::getSupportedLanguagesKeys();
    }

}

```

```php
// resources/views/sitemap/xml.blade.php
{!! '<'.'?'.'xml version="1.0" encoding="UTF-8"?>' !!}

<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9" xmlns:xhtml="http://www.w3.org/1999/xhtml">

@foreach($routes as $route)
    <url>
        <loc>{{ $route['loc'] }}</loc>
        @foreach($route['alternates'] as $locale => $url)
            <xhtml:link rel="alternate" hreflang="{{ $locale }}" href="{{ $url }}" />
        @endforeach
        <priority>{{ $route['priority'] }}</priority>
    </url>
@endforeach
</urlset>
```

**Comment by nmpribeiro** [comment]
Hi @jaumesala that's awesome... I've just made my package, still not published.
But instantly I got a barrier: what if each user places routes on his own desired place? And how about 'auth.email.verify' routes and routes that the user doesn't translate? Need to think about, but we'll need to ask user for a config I suppose.

In my case, to translate routes, I have routes/web.php . We can argue this isn't needed, but nevertheless, it's a week link to make an 'all general purpose solution'.
